src/mountinfo.rs

```diff
@@ -0,0 +1,418 @@
+    let root = try!(parse_path(&mut cols, row, row_num));
+    let mount_point = try!(parse_path(&mut cols, row, row_num));
+    let mount_options = try!(parse_os_str(&mut cols, row, row_num));
+    let mut optional_fields = vec!();
+    let mut opt_field = try!(parse_os_str(&mut cols, row, row_num));
+    while opt_field != OsStr::new("-") {
+        optional_fields.push(opt_field.clone());
+        opt_field = try!(parse_os_str(&mut cols, row, row_num));
+    }
+    let fstype = try!(parse_os_str(&mut cols, &row, row_num));
+    let mount_source = try!(parse_os_str(&mut cols, &row, row_num));
+    let super_options = try!(parse_os_str(&mut cols, &row, row_num));
+    Ok(MountPoint {
+        mount_id: mount_id,
+        parent_id: parent_id,
+        major: major,
+        minor: minor,
+        root: root,
+        mount_point: mount_point,
+        mount_options: mount_options,
+        optional_fields: optional_fields,
+        fstype: fstype,
+        mount_source: mount_source,
+        super_options: super_options,
+    })
+}
+
+fn is_comment_line(row: &[u8]) -> bool {
+    if row.is_empty() {
+        return true;
+    }
+    for c in row {
+        if *c == b' ' || *c == b'\t' {
+            continue;
+        }
+        if *c == b'#' {
+            return true;
+        }
+        return false;
+    }
+    return false;
+}
+
+fn strip_line(row: &[u8]) -> &[u8] {
+    let mut row = row;
+    while row.ends_with(&[b'\r']) {
+        row = &row[..row.len() - 1]
+    }
+    row
+}
+
+fn parse_os_str(columns: &mut Iterator<Item=&[u8]>, row: &[u8], row_num: usize)
+    -> Result<OsString, ParseError>
+{
+    let bytes = try!(columns.next()
+        .ok_or_else(|| ParseError(
+            format!("Expected more fields"),
+            row_num, String::from_utf8_lossy(row).into_owned())));
+    let mut value = Cow::Borrowed(bytes);
+    unescape_octals(&mut value);
+    Ok(OsString::from_vec(value.into_owned()))
+}
+
+fn parse_int(columns: &mut Iterator<Item=&[u8]>, row: &[u8], row_num: usize)
+    -> Result<c_ulong, ParseError>
+{
+    let col = try!(columns.next()
+        .ok_or_else(|| ParseError(
+            format!("Expected more fields"),
+            row_num, String::from_utf8_lossy(row).into_owned()))
+        .map(|v| String::from_utf8_lossy(v)));
+    col.parse::<c_ulong>()
+        .map_err(|_| ParseError(
+            format!("Cannot parse integer: {:?}", col),
+            row_num, String::from_utf8_lossy(row).into_owned()))
+}
+
+fn parse_path(columns: &mut Iterator<Item=&[u8]>, row: &[u8], row_num: usize)
+    -> Result<PathBuf, ParseError>
+{
+    Ok(PathBuf::from(try!(parse_os_str(columns, row, row_num))))
+}
+
+fn unescape_octals(v: &mut Cow<[u8]>) {
+    let mut i = 0;
+    while i < v.len() {
+        if v[i] == b'\\' && v[i..].len() >= 4
+            && is_oct(v[i+1]) && is_oct(v[i+2]) && is_oct(v[i+3])
+        {
+            let t = v.to_mut().split_off(i);
```

Comment:
Wow. You allocate a new vector here, for every single escape character.

I mean the complexity of this algo is `(number of characters in string) x (number of escapes)`